clippy_lints/src/hash_collision.rs

```diff
@@ -0,0 +1,91 @@
+use clippy_utils::diagnostics::span_lint_and_note;
+use clippy_utils::last_path_segment;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_hir::ExprKind;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// When two items are inserted into a `HashMap` with the same key,
+    /// the second item will overwrite the first item.
+    ///
+    /// ### Why is this bad?
+    /// This can lead to data loss.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// # use std::collections::HashMap;
+    /// let example = HashMap::from([(5, 1), (5, 2)]);
+    /// ```
+    #[clippy::version = "1.79.0"]
+    pub HASH_COLLISION,
```

Comment:
"Hash collision" to me means different keys with the same hash, but this lint specifically looks for same keys.
(A hash collision also isn't bad in and of itself and won't lead to overwriting entries, a hashmap can deal with that)

Should this be renamed from `hash_collision` to something else? What about `duplicate_map_keys`?

Reply:
Thanks, I should have caught that. `duplicate_map_keys` is a good name, I've changed it in the latest commit.